src/com/company/Shop2.java

```diff
@@ -1,59 +1,81 @@
 package com.company;
 
 import java.util.HashMap;
-import java.util.List;
+
 import java.util.Map;
+import java.util.Objects;
 
 public class Shop2 implements Shop {
-    private Map<Product, Integer> mapOfProducts = new HashMap();
-    private int counter;
-    private double doubleCounter;
+    private Map<Product, Integer> storageMap = new HashMap();
+    private Map<Long, Product> mapOfId = new HashMap<>();
+    private Map<Product, Integer> mapOfShoppingCart = new HashMap<>();
+    private int intCounter;
```

Comment:
what is counter? why it's a field? if you use something as a temporal counter, make it a local variable